panic in the TestAlterAddConstraintStateChange3

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

```
panic: runtime error: index out of range [0] with length 0
goroutine 80540 [running]:
github.com/pingcap/tidb/pkg/ddl_test.TestAlterAddConstraintStateChange3.func1(0xc011fc5e00)
	pkg/ddl/constraint_test.go:235 +0x5af
github.com/pingcap/tidb/pkg/ddl/util/callback.(*TestDDLCallback).OnJobUpdated(0xc013e23e30, 0xc011fc5e00?)
	pkg/ddl/util/callback/callback.go:124 +0x1dd
github.com/pingcap/tidb/pkg/ddl.(*ddl).delivery2worker.func1()
	pkg/ddl/job_table.go:439 +0x171b
github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1()
	pkg/util/wait_group_wrapper.go:157 +0x87
created by github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run in goroutine 79460
	pkg/util/wait_group_wrapper.go:155 +0xe5 
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/ddl/constraint_test.go
+++ b/pkg/ddl/constraint_test.go
@@ -130,15 +130,11 @@ func TestAlterAddConstraintStateChange1(t *testing.T) {
 	tk1.MustExec("use test")
 	tk1.MustExec("insert into t values(12)")
 
-	var checkErr error
 	d := dom.DDL()
 	originalCallback := d.GetHook()
 	callback := &callback.TestDDLCallback{}
 	// StatNone -> StateWriteOnly
 	onJobUpdatedExportedFunc1 := func(job *model.Job) {
-		if checkErr != nil {
-			return
-		}
 		originalCallback.OnChanged(nil)
 		if job.SchemaState == model.StateWriteOnly {
 			// set constraint state
@@ -174,15 +170,11 @@ func TestAlterAddConstraintStateChange2(t *testing.T) {
 	tk1.MustExec("use test")
 	tk1.MustExec("insert into t values(12)")
 
-	var checkErr error
 	d := dom.DDL()
 	originalCallback := d.GetHook()
 	callback := &callback.TestDDLCallback{}
 	// StateWriteOnly -> StateWriteReorganization
 	onJobUpdatedExportedFunc2 := func(job *model.Job) {
-		if checkErr != nil {
-			return
-		}
 		originalCallback.OnChanged(nil)
 		if job.SchemaState == model.StateWriteReorganization {
 			// set constraint state
@@ -217,13 +209,13 @@ func TestAlterAddConstraintStateChange3(t *testing.T) {
 	tk1.MustExec("use test")
 	tk1.MustExec("insert into t values(12)")
 
-	var checkErr error
+	addCheckDone := false
 	d := dom.DDL()
 	originalCallback := d.GetHook()
 	callback := &callback.TestDDLCallback{}
 	// StateWriteReorganization -> StatePublic
 	onJobUpdatedExportedFunc3 := func(job *model.Job) {
-		if checkErr != nil {
+		if job.Type != model.ActionAddCheckConstraint || job.TableName != "t" {
 			return
 		}
 		originalCallback.OnChanged(nil)
@@ -239,13 +231,19 @@ func TestAlterAddConstraintStateChange3(t *testing.T) {
 			// recover
 			tableCommon.Constraints[0].State = model.StatePublic
 			tableCommon.WritableConstraint()
+			addCheckDone = true
 		}
 	}
 	callback.OnJobUpdatedExported.Store(&onJobUpdatedExportedFunc3)
 	d.SetHook(callback)
 	tk.MustExec("alter table t add constraint c3 check ( a > 10)")
 	// Issue TiDB#48123.
-	time.Sleep(50 * time.Millisecond)
+	for i := 0; i <= 100; i++ {
+		if addCheckDone {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
 	tk.MustQuery("select * from t").Check(testkit.Rows("12"))
 	tk.MustQuery("show create table t").Check(testkit.Rows("t CREATE TABLE `t` (\n  `a` int(11) DEFAULT NULL,\nCONSTRAINT `c3` CHECK ((`a` > 10))\n) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin"))
 }
@@ -261,15 +259,11 @@ func TestAlterEnforcedConstraintStateChange(t *testing.T) {
 	tk1.MustExec("use test")
 	tk1.MustExec("insert into t values(12)")
 
-	var checkErr error
 	d := dom.DDL()
 	originalCallback := d.GetHook()
 	callback := &callback.TestDDLCallback{}
 	// StateWriteReorganization -> StatePublic
 	onJobUpdatedExportedFunc3 := func(job *model.Job) {
-		if checkErr != nil {
-			return
-		}
 		originalCallback.OnChanged(nil)
 		if job.SchemaState == model.StateWriteReorganization {
 			// set constraint state

--- a/pkg/ddl/index_change_test.go
+++ b/pkg/ddl/index_change_test.go
@@ -55,6 +55,9 @@ func TestIndexChange(t *testing.T) {
 		publicTable     table.Table
 	)
 	onJobUpdatedExportedFunc := func(job *model.Job) {
+		if job.Type != model.ActionAddIndex || job.TableName != "t" {
+			return
+		}
 		if job.SchemaState == prevState {
 			return
 		}
@@ -87,11 +90,11 @@ func TestIndexChange(t *testing.T) {
 	// We need to make sure onJobUpdated is called in the first hook.
 	// After testCreateIndex(), onJobUpdated() may not be called when job.state is Sync.
 	// If we skip this check, prevState may wrongly set to StatePublic.
-	for i := 0; i <= 10; i++ {
+	for i := 0; i <= 100; i++ {
 		if addIndexDone {
 			break
 		}
-		time.Sleep(50 * time.Millisecond)
+		time.Sleep(10 * time.Millisecond)
 	}
 	v := getSchemaVer(t, tk.Session())
 	checkHistoryJobArgs(t, tk.Session(), jobID.Load(), &historyJobArgs{ver: v, tbl: publicTable.Meta()})

